probe-rs/src/probe/daplink/commands/mod.rs

```diff
@@ -71,16 +82,54 @@ impl DAPLinkDevice {
                 handle,
                 out_ep,
                 in_ep: _,
+                swo_ep: _,
             } => {
                 let timeout = Duration::from_millis(100);
                 // Skip first byte as it's set to 0 for HID transfers
                 Ok(handle.write_bulk(*out_ep, &buf[1..], timeout)?)
             }
         }
     }
+
+    /// Check if SWO streaming is supported by this device.
+    pub(super) fn swo_streaming_supported(&self) -> bool {
+        match self {
+            DAPLinkDevice::V1(_) => false,
+            DAPLinkDevice::V2 {
+                handle: _,
+                out_ep: _,
+                in_ep: _,
+                swo_ep,
+            } => swo_ep.is_some(),
+        }
+    }
+
+    /// Read into `buf` from the SWO streaming endpoint.
+    ///
+    /// Returns SWOModeNotAvailable if this device does not support SWO streaming.
+    ///
+    /// On timeout, returns Ok(0).
+    pub(super) fn read_swo_stream(&self, buf: &mut [u8], timeout: Duration) -> Result<usize> {
+        match self {
+            DAPLinkDevice::V1(_) => Err(CmsisDapError::SWOModeNotAvailable.into()),
+            DAPLinkDevice::V2 {
+                handle,
+                out_ep: _,
+                in_ep: _,
+                swo_ep,
+            } => match swo_ep {
+                Some(ep) => match handle.read_bulk(*ep, buf, timeout) {
+                    Ok(n) => Ok(n),
+                    Err(rusb::Error::Timeout) => Ok(0),
```

Comment:
I think this choice is generally okay, but what if something goes wrong and we timeout over and over again, then we just hang here without any indication, no?

Reply:
If we timeout we'll return from this method all the way to the user application which called `read_swo`, saying we had no bytes available. The user application can then re-read until bytes become available. If there was some error that means no more data will ever arrive, we should get a different error type from rusb and send that up. If we do just keep getting a timeout forever that's the same as if there was no data received from the target forever, i.e. the user application should either just also keep waiting for data or eventually give up/the user terminates it.

Reply:
Okay, sounds good to me :)